### ISSUE TITLE
[Bug]: UnitTest inconsistency with XUnit using Realm

### ISSUE THREAD

**Issue by PawKanarek** [realm/realm-dotnet#2965]
### What happened?

My test are no longer consistent and they randomly crash.  I've managed to reproduce this issue on very simple project (Snippet is below). Sometimes there is an error ```134 (/Users/runner/work/realm-dotnet/realm-dotnet/wrappers/realm-core/src/realm/object-store/impl/realm_coordinator.cpp:1250: [realm-core-12.1.0] Assertion failed: !realm.is_in_transaction()```
sometimes error is simple `exit code is 139` 

Test are failing on my local machine (MacOS), and on Jenkins CI (Windows). 

using terminal command `dotnet test TestProject1.csproj` also randomly crashes. So it's not fault of the JetBrains Rider. 

Do you have any idea what is happening? 

Additional proof: 

https://user-images.githubusercontent.com/15096514/175337711-69bffc1b-d41f-433d-848e-2e87ba0c2f28.mov

### Repro steps

Launch many times below project

### Version

netcoreapp3.1

### What SDK flavour are you using?

Local Database only

### What type of application is this?

Xamarin

### Client OS and version

MacOS, Windows, Xamarin, Tests

### Code snippets

```csharp
using System;
using System.IO;
using System.Linq;
using System.Runtime.CompilerServices;
using System.Threading.Tasks;
using Realms;
using Xunit;
using Xunit.Abstractions;

namespace TestProject1
{
    public class UnitTest1
    {
        [Fact]
        public async Task TestAsync()
        {
            var realmConfig = ArrangeTest();
            using var realm0 = await Realm.GetInstanceAsync(realmConfig);
            for (int i = 0; i < 10; i++) // test some async stuff
            {
                await ViewModelMock.SimulateHeavyComputationAsync(realmConfig).ConfigureAwait(false);
            }

            using var realm = await Realm.GetInstanceAsync(realmConfig);
            var count = realm.All<RealmObj>().ToList().Count;
            Assert.Equal(count, 100);
        }

        [Fact]
        public void Test()
        {
            var realmConfig = ArrangeTest();
            using var realm0 = Realm.GetInstance(realmConfig);

            for (int i = 0; i < 10; i++) // test some stuff
            {
                ViewModelMock.SimulateHeavyComputation(realmConfig);
            }

            // assert with disposing 
            using var realm = Realm.GetInstance(realmConfig);
            var count = realm.All<RealmObj>().ToList().Count;
            Assert.Equal(count, 100);
        }

        private RealmConfiguration ArrangeTest([CallerMemberName] string name = null)
        {
            var realmConfig = new RealmConfiguration($"{name}.realm");
            File.Delete(realmConfig.DatabasePath); // always reset realm
            return realmConfig;
        }
    }

    public static class ViewModelMock
    {
        public static async Task SimulateHeavyComputationAsync(RealmConfiguration config)
        {
            await Task.Run(() =>
            {
                for (int i = 0; i < 10; i++)
                {
                    using var realm = Realm.GetInstance(config);
                    using var transaction = realm.BeginWrite();
                    realm.Add(new RealmObj());
                    transaction.Commit();
                }
            }).ConfigureAwait(false);
        }

        public static void SimulateHeavyComputation(RealmConfiguration config)
        {
            for (int i = 0; i < 10; i++)
            {
                using var realm = Realm.GetInstance(config);
                using var transaction = realm.BeginWrite();
                realm.Add(new RealmObj());
                transaction.Commit();
            }
        }
    }

    public class RealmObj : RealmObject
    {
        [PrimaryKey] private Guid Id { get; set; } = Guid.NewGuid();
    }
}
```

```csproj
<Project Sdk="Microsoft.NET.Sdk">

    <PropertyGroup>
        <TargetFramework>netcoreapp3.1</TargetFramework>

        <IsPackable>false</IsPackable>
    </PropertyGroup>

    <ItemGroup>
        <PackageReference Include="Fody" Version="6.6.3">
          <PrivateAssets>all</PrivateAssets>
          <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
        </PackageReference>
        <PackageReference Include="Microsoft.NET.Test.Sdk" Version="16.5.0" />
        <PackageReference Include="Nito.AsyncEx.Context" Version="5.1.2" />
        <PackageReference Include="Realm" Version="10.14.0" />
        <PackageReference Include="Realm.Fody" Version="10.14.0">
          <PrivateAssets>all</PrivateAssets>
          <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
        </PackageReference>
        <PackageReference Include="xunit" Version="2.4.1" />
        <PackageReference Include="xunit.runner.visualstudio" Version="2.4.3" />
        <PackageReference Include="coverlet.collector" Version="1.2.0" />
    </ItemGroup>

</Project>

```

### Stacktrace of the exception/crash you're getting

```shell
17:17:30.522 |I| Launch created for session 'UnitTest1' (26985d17-f486-4d2f-a3c8-5b90af497c71)
17:17:30.523 |T| Launch: ed7c88be-23a8-4cbf-b090-d086a01e71db >> RunHostController.AfterLaunchStarted
17:17:30.523 |T| Launch: ed7c88be-23a8-4cbf-b090-d086a01e71db << RunHostController.AfterLaunchStarted
17:17:30.523 |I| BuildStage started
17:17:30.523 |I| Build policy: Automatic
17:17:31.109 |I| 0 projects built
17:17:31.109 |V| 
17:17:31.109 |I| BuildStage completed
17:17:31.109 |I| RefreshProjectPropertiesStage started
17:17:31.109 |I| RefreshProjectPropertiesStage completed
17:17:31.109 |I| DiscoveryStage started
17:17:31.109 |I| Test exploration required for 0 projects
17:17:31.109 |V| 
17:17:31.110 |I| DiscoveryStage completed
17:17:31.110 |I| BuildPipelineStage started
17:17:31.110 |I| BuildPipelineStage completed
17:17:31.110 |I| ApplyRulesStage started
17:17:31.110 |T| Removing elements marked with ignored categories, if any...
17:17:31.110 |T| Removing empty parent elements, if any...
17:17:31.110 |V| Running elements from criterion: {ConjunctiveCriterion: {TestAncestorCriterion: TestProject1.UnitTest1;E872A650-4B7C-449F-9EFD-958F12B941E0;.NETCoreApp,Version=v3.1;xUnit}, {ProjectCriterion: /Users/pkanarek/RiderProjects/TestProject1/TestProject1}, {NamespaceCriterion: TestProject1}}
17:17:31.110 |I| Got 3 elements (1 explicit) to run
17:17:31.110 |V| Marking 3 elements as Pending
17:17:31.111 |V| Marking elements as Pending completed
17:17:31.111 |V| Distributing 3 elements into runs
17:17:31.112 |V| Distributing elements completed
17:17:31.112 |I| Created 1 unit-test runs
17:17:31.112 |I| ApplyRulesStage completed
17:17:31.112 |I| RunTestsStage started
17:17:31.112 |I| Running up to 1 unit-test runs in parallel
17:17:31.112 |I| Run: d2a590f5-2c5f-4887-a352-e9820aa3196d - Starting
17:17:31.112 |V|     Provider: xUnit.net
    Target Framework: netcoreapp3.1
    Strategy: XUnitTestRunnerRunStrategy
    Runtime Environment: TestRunnerRuntimeEnvironment
  Host: JetBrains.ReSharper.UnitTestFramework.Execution.TestRunner.DefaultTestRunnerHost
  Project: TestProject1
  TargetPlatform: X64
  TargetFrameworkId: .NETCoreApp,Version=v3.1
  RunAsTargetFrameworkId: .NETCoreApp,Version=v3.1
  IsUnmanaged: False
17:17:31.112 |T| Run: d2a590f5-2c5f-4887-a352-e9820aa3196d >> RunHostController.PrepareForRun
17:17:31.112 |T| Run: d2a590f5-2c5f-4887-a352-e9820aa3196d << RunHostController.PrepareForRun
17:17:31.116 |V| RunSettings: 
<RunSettings><RunConfiguration><ResultsDirectory>/Users/pkanarek/RiderProjects/TestProject1/TestResults</ResultsDirectory><SolutionDirectory>/Users/pkanarek/RiderProjects/TestProject1</SolutionDirectory><TargetPlatform>X64</TargetPlatform></RunConfiguration></RunSettings>
17:17:31.117 |V| [ProcessStartInfoPatchResult]
  Original request: DotNetCoreRequest
  Success: True
  ErrorMessage: <Empty>
  Original StartInfo: FileName = 'exec', Arguments = '  --runtimeConfig /Users/pkanarek/RiderProjects/TestProject1/TestProject1/bin/Debug/netcoreapp3.1/TestProject1.runtimeconfig.json --depsfile /Users/pkanarek/RiderProjects/TestProject1/TestProject1/bin/Debug/netcoreapp3.1/TestProject1.deps.json  --additional-deps /Applications/Rider.app/Contents/lib/ReSharperHost/TestRunner/netcoreapp3.0/ReSharperTestRunner.deps.json --additionalprobingpath /Applications/Rider.app/Contents/lib/ReSharperHost/TestRunner/netcoreapp3.0 /Applications/Rider.app/Contents/lib/ReSharperHost/TestRunner/netcoreapp3.0/ReSharperTestRunner.dll --parentProcessId 18609 -p 53551 -r bb195a57-6ab6-47db-a29b-d65e832ef2bb', WorkDir = '/Users/pkanarek/RiderProjects/TestProject1/TestProject1/bin/Debug/netcoreapp3.1'
  Patched  StartInfo: FileName = '/usr/local/share/dotnet/dotnet', Arguments = '"exec"   --runtimeConfig /Users/pkanarek/RiderProjects/TestProject1/TestProject1/bin/Debug/netcoreapp3.1/TestProject1.runtimeconfig.json --depsfile /Users/pkanarek/RiderProjects/TestProject1/TestProject1/bin/Debug/netcoreapp3.1/TestProject1.deps.json  --additional-deps /Applications/Rider.app/Contents/lib/ReSharperHost/TestRunner/netcoreapp3.0/ReSharperTestRunner.deps.json --additionalprobingpath /Applications/Rider.app/Contents/lib/ReSharperHost/TestRunner/netcoreapp3.0 /Applications/Rider.app/Contents/lib/ReSharperHost/TestRunner/netcoreapp3.0/ReSharperTestRunner.dll --parentProcessId 18609 -p 53551 -r bb195a57-6ab6-47db-a29b-d65e832ef2bb', WorkDir = '/Users/pkanarek/RiderProjects/TestProject1/TestProject1/bin/Debug/netcoreapp3.1'

17:17:31.117 |I| Starting process: /usr/local/share/dotnet/dotnet "exec"   --runtimeConfig /Users/pkanarek/RiderProjects/TestProject1/TestProject1/bin/Debug/netcoreapp3.1/TestProject1.runtimeconfig.json --depsfile /Users/pkanarek/RiderProjects/TestProject1/TestProject1/bin/Debug/netcoreapp3.1/TestProject1.deps.json  --additional-deps /Applications/Rider.app/Contents/lib/ReSharperHost/TestRunner/netcoreapp3.0/ReSharperTestRunner.deps.json --additionalprobingpath /Applications/Rider.app/Contents/lib/ReSharperHost/TestRunner/netcoreapp3.0 /Applications/Rider.app/Contents/lib/ReSharperHost/TestRunner/netcoreapp3.0/ReSharperTestRunner.dll --parentProcessId 18609 -p 53551 -r bb195a57-6ab6-47db-a29b-d65e832ef2bb
17:17:31.165 |I| Started process with pid 27973
17:17:31.578 |V| Test runner agent <bb195a57-6ab6-47db-a29b-d65e832ef2bb> connected
17:17:31.580 |V| Registering remote tasks...
17:17:31.581 |V| Registering message handlers from 'JetBrains.ReSharper.UnitTestFramework.Execution.TestRunner.DefaultTestRunnerHost'...
17:17:31.584 |V| Registering message handlers from 'JetBrains.ReSharper.UnitTesting.MSTest.Provider.DotNetCore.Proprietary.MsTestTestRunnerOrchestrator'...
17:17:31.584 |V| Registering message handlers from 'JetBrains.ReSharper.UnitTesting.VsTest.Provider.VsTestTestRunnerOrchestrator'...
17:17:31.584 |V| Registering message handlers from 'JetBrains.ReSharper.UnitTestProvider.nUnit.v30.NUnitTestRunnerOrchestrator'...
17:17:31.584 |V| Registering message handlers from 'JetBrains.ReSharper.UnitTestProvider.nUnit.v26.NUnitTestRunnerOrchestrator'...
17:17:31.584 |V| Registering message handlers from 'JetBrains.ReSharper.UnitTestProvider.Xunit.DotNetCore.Proprietary.XUnitTestRunnerOrchestrator'...
17:17:31.601 |V| TestRunner: JetBrains.ReSharper.TestRunner.Implementation.TestAdapterLoadContextFactoryFactory Registered protocol entities: 8 scalars and 1 serializers 
17:17:31.616 |I| TestRunner: JetBrains.ReSharper.TestRunner.Isolation.ShadowCopyBehavior+NoShadowCopyBehavior Shadow-copy will be skipped as it is disabled 
17:17:31.619 |V| TestRunner: JetBrains.ReSharper.TestRunner.Implementation.DefaultTestAdapterLoadContextFactory Initializing test adapter load context: NoIsolation = True 
17:17:31.622 |V| TestRunner: AssemblyLoader(Application) Added probing path '/Users/pkanarek/RiderProjects/TestProject1/TestProject1/bin/Debug/netcoreapp3.1' 
17:17:31.627 |V| TestRunner: AssemblyLoader(Application) Added resolver relative to '/Users/pkanarek/RiderProjects/TestProject1/TestProject1/bin/Debug/netcoreapp3.1/TestProject1.dll' 
17:17:31.628 |V| TestRunner: AssemblyLoader(Application) Added probing path '/Applications/Rider.app/Contents/lib/ReSharperHost/TestRunner/Adapters/XUnit/netstandard2.0' 
17:17:31.629 |V| TestRunner: AssemblyLoader(Application) Added resolver relative to '/Applications/Rider.app/Contents/lib/ReSharperHost/TestRunner/Adapters/XUnit/netstandard2.0/JetBrains.ReSharper.TestRunner.Adapters.XUnit.dll' 
17:17:31.630 |V| TestRunner: JetBrains.ReSharper.TestRunner.Implementation.TestAdapterLoadContextWrapper Initialized test executor: JetBrains.ReSharper.TestRunner.Adapters.XUnit.XUnitRunner, JetBrains.ReSharper.TestRunner.Adapters.XUnit, Version=2.9.1.29, Culture=neutral, PublicKeyToken=5c492ec4f3eccde3 
17:17:31.649 |I| TestRunner: JetBrains.ReSharper.TestRunner.Adapters.XUnit.XUnitRunner Execution started 
17:17:31.654 |I| TestRunner: JetBrains.ReSharper.TestRunner.Adapters.XUnit.Discoverer Discovering tests from /Users/pkanarek/RiderProjects/TestProject1/TestProject1/bin/Debug/netcoreapp3.1/TestProject1.dll 
17:17:31.721 |V| TestRunner: JetBrains.ReSharper.TestRunner.Adapters.XUnit.Discoverer Sending 3 discovered elements to server... 
17:17:31.725 |V| Run: d2a590f5-2c5f-4887-a352-e9820aa3196d - Discovery result processing started
17:17:31.725 |V| Run: d2a590f5-2c5f-4887-a352-e9820aa3196d - Discovery result processing finished (+0 -0 ~3)
17:17:31.726 |I| TestRunner: JetBrains.ReSharper.TestRunner.Adapters.XUnit.Discoverer Discovery completed 
17:17:31.931 |I| Process /usr/local/share/dotnet/dotnet:27973 has exited with code (134)
17:17:31.931 |I| Output stream: /Users/runner/work/realm-dotnet/realm-dotnet/wrappers/realm-core/src/realm/object-store/impl/realm_coordinator.cpp:1250: [realm-core-12.1.0] Assertion failed: !realm.is_in_transaction()
0   librealm-wrappers.dylib             0x000000010a9dbc91 _ZN5realm4util9terminateEPKcS2_lOSt16initializer_listINS0_9PrintableEE + 369
1   librealm-wrappers.dylib             0x000000010a608824 _ZN5realm5_impl16RealmCoordinator23process_available_asyncERNS_5RealmE + 1092
2   librealm-wrappers.dylib             0x000000010a697883 _ZN5realm5Realm6notifyEv + 275
3   librealm-wrappers.dylib             0x000000010a621f11 _ZN5realm4util14UniqueFunctionIFvvEE12SpecificImplIZNS_5_impl17WeakRealmNotifier6notifyEvE3$_0E4callEv + 49
4   librealm-wrappers.dylib             0x000000010a59295f realm_scheduler_invoke_function + 31
5   ???                                 0x0000000117cb02bb 0x0 + 4694147771
6   ???                                 0x0000000117cb0218 0x0 + 4694147608
7   ???                                 0x0000000117cb010b 0x0 + 4694147339
8   ???                                 0x000000011782e5ed 0x0 + 4689421805
9   ???                                 0x000000011782e583 0x0 + 4689421699
10  ???                                 0x0000000115acf591 0x0 + 4658623889
11  ???                                 0x000000011782e502 0x0 + 4689421570
12  ???                                 0x000000011782bdea 0x0 + 4689411562
13  ???                                 0x00000001178116a2 0x0 + 4689303202
14  ???                                 0x0000000115ae5203 0x0 + 4658713091
15  ???                                 0x0000000115aeabd2 0x0 + 4658736082
16  ???                                 0x0000000115acf591 0x0 + 4658623889
17  ???                                 0x0000000115ae5032 0x0 + 4658712626
18  ???                                 0x0000000115ae4e68 0x0 + 4658712168
19  ???                                 0x0000000115af1deb 0x0 + 4658765291
20  ???                                 0x0000000115acf591 0x0 + 4658623889
21  ???                                 0x0000000115ac5d91 0x0 + 4658584977
22  libcoreclr.dylib                    0x00000001074fe03d CallDescrWorkerInternal + 124
23  libcoreclr.dylib                    0x000000010736081a _ZN18MethodDescCallSite16CallTargetWorkerEPKmPmi + 1530
24  libcoreclr.dylib                    0x00000001073740ef _ZN12ThreadNative20KickOffThread_WorkerEPv + 447
25  libcoreclr.dylib                    0x00000001073245f8 _ZL31ManagedThreadBase_DispatchOuterP22ManagedThreadCallState + 296
26  libcoreclr.dylib                    0x0000000107324bd0 _ZN17ManagedThreadBase7KickOffEPFvPvES0_ + 32
27  libcoreclr.dylib                    0x000000010737431b _ZN12ThreadNative13KickOffThreadEPv + 347
28  libcoreclr.dylib                    0x00000001071ad881 _ZN7CorUnix10CPalThread11ThreadEntryEPv + 417
29  libsystem_pthread.dylib             0x00007ff8178564e1 _pthread_start + 125
30  libsystem_pthread.dylib             0x00007ff817851f6b thread_start + 15!!! IMPORTANT: Please report this at https://github.com/realm/realm-core/issues/new/choose


17:17:32.033 |E| Process /usr/local/share/dotnet/dotnet:27973 exited with code '134': /Users/runner/work/realm-dotnet/realm-dotnet/wrappers/realm-core/src/realm/object-store/impl/realm_coordinator.cpp:1250: [realm-core-12.1.0] Assertion failed: !realm.is_in_transaction() 0 librealm-wrappers.dylib 0x000000010a9dbc91 _ZN5realm4util9terminateEPKcS2_lOSt16initializer_listINS0_9PrintableEE + 369 1 librealm-wrappers.dylib 0x000000010a608824 _ZN5realm5_impl16RealmCoordinator23process_available_asyncERNS_5RealmE + 1092 2 librealm-wrappers.dylib 0x000000010a697883 _ZN5realm5Realm6notifyEv + 275 3 librealm-wrappers.dylib 0x000000010a621f11 _ZN5realm4util14UniqueFunctionIFvvEE12SpecificImplIZNS_5_impl17WeakRealmNotifier6notifyEvE3$_0E4callEv + 49 4 librealm-wrappers.dylib 0x000000010a59295f realm_scheduler_invoke_function + 31 5 ??? 0x0000000117cb02bb 0x0 + 4694147771 6 ??? 0x0000000117cb0218 0x0 + 4694147608 7 ??? 0x0000000117cb010b 0x0 + 4694147339 8 ??? 0x000000011782e5ed 0x0 + 4689421805 9 ??? 0x000000011782e583 0x0 + 4689421699 10 ??? 0x0000000115acf591 0x0 + 4658623889 11 ??? 0x000000011782e502 0x0 + 4689421570 12 ??? 0x000000011782bdea 0x0 + 4689411562 13 ??? 0x00000001178116a2 0x0 + 4689303202 14 ??? 0x0000000115ae5203 0x0 + 4658713091 15 ??? 0x0000000115aeabd2 0x0 + 4658736082 16 ??? 0x0000000115acf591 0x0 + 4658623889 17 ??? 0x0000000115ae5032 0x0 + 4658712626 18 ??? 0x0000000115ae4e68 0x0 + 4658712168 19 ??? 0x0000000115af1deb 0x0 + 4658765291 20 ??? 0x0000000115acf591 0x0 + 4658623889 21 ??? 0x0000000115ac5d91 0x0 + 4658584977 22 libcoreclr.dylib 0x00000001074fe03d CallDescrWorkerInternal + 124 23 libcoreclr.dylib 0x000000010736081a _ZN18MethodDescCallSite16CallTargetWorkerEPKmPmi + 1530 24 libcoreclr.dylib 0x00000001073740ef _ZN12ThreadNative20KickOffThread_WorkerEPv + 447 25 libcoreclr.dylib 0x00000001073245f8 _ZL31ManagedThreadBase_DispatchOuterP22ManagedThreadCallState + 296 26 libcoreclr.dylib 0x0000000107324bd0 _ZN17ManagedThreadBase7KickOffEPFvPvES0_ + 32 27 libcoreclr.dylib 0x000000010737431b _ZN12ThreadNative13KickOffThreadEPv + 347 28 libcoreclr.dylib 0x00000001071ad881 _ZN7CorUnix10CPalThread11ThreadEntryEPv + 417 29 libsystem_pthread.dylib 0x00007ff8178564e1 _pthread_start + 125 30 libsystem_pthread.dylib 0x00007ff817851f6b thread_start + 15!!! IMPORTANT: Please report this at https://github.com/realm/realm-core/issues/new/choose

--- EXCEPTION #1/1 [LoggerException]
Message = “
  Process /usr/local/share/dotnet/dotnet:27973 exited with code '134':
  /Users/runner/work/realm-dotnet/realm-dotnet/wrappers/realm-core/src/realm/object-store/impl/realm_coordinator.cpp:1250: [realm-core-12.1.0] Assertion failed: !realm.is_in_transaction()
  0   librealm-wrappers.dylib             0x000000010a9dbc91 _ZN5realm4util9terminateEPKcS2_lOSt16initializer_listINS0_9PrintableEE + 369
  1   librealm-wrappers.dylib             0x000000010a608824 _ZN5realm5_impl16RealmCoordinator23process_available_asyncERNS_5RealmE + 1092
  2   librealm-wrappers.dylib             0x000000010a697883 _ZN5realm5Realm6notifyEv + 275
  3   librealm-wrappers.dylib             0x000000010a621f11 _ZN5realm4util14UniqueFunctionIFvvEE12SpecificImplIZNS_5_impl17WeakRealmNotifier6notifyEvE3$_0E4callEv + 49
  4   librealm-wrappers.dylib             0x000000010a59295f realm_scheduler_invoke_function + 31
  5   ???                                 0x0000000117cb02bb 0x0 + 4694147771
  6   ???                                 0x0000000117cb0218 0x0 + 4694147608
  7   ???                                 0x0000000117cb010b 0x0 + 4694147339
  8   ???                                 0x000000011782e5ed 0x0 + 4689421805
  9   ???                                 0x000000011782e583 0x0 + 4689421699
  10  ???                                 0x0000000115acf591 0x0 + 4658623889
  11  ???                                 0x000000011782e502 0x0 + 4689421570
  12  ???                                 0x000000011782bdea 0x0 + 4689411562
  13  ???                                 0x00000001178116a2 0x0 + 4689303202
  14  ???                                 0x0000000115ae5203 0x0 + 4658713091
  15  ???                                 0x0000000115aeabd2 0x0 + 4658736082
  16  ???                                 0x0000000115acf591 0x0 + 4658623889
  17  ???                                 0x0000000115ae5032 0x0 + 4658712626
  18  ???                                 0x0000000115ae4e68 0x0 + 4658712168
  19  ???                                 0x0000000115af1deb 0x0 + 4658765291
  20  ???                                 0x0000000115acf591 0x0 + 4658623889
  21  ???                                 0x0000000115ac5d91 0x0 + 4658584977
  22  libcoreclr.dylib                    0x00000001074fe03d CallDescrWorkerInternal + 124
  23  libcoreclr.dylib                    0x000000010736081a _ZN18MethodDescCallSite16CallTargetWorkerEPKmPmi + 1530
  24  libcoreclr.dylib                    0x00000001073740ef _ZN12ThreadNative20KickOffThread_WorkerEPv + 447
  25  libcoreclr.dylib                    0x00000001073245f8 _ZL31ManagedThreadBase_DispatchOuterP22ManagedThreadCallState + 296
  26  libcoreclr.dylib                    0x0000000107324bd0 _ZN17ManagedThreadBase7KickOffEPFvPvES0_ + 32
  27  libcoreclr.dylib                    0x000000010737431b _ZN12ThreadNative13KickOffThreadEPv + 347
  28  libcoreclr.dylib                    0x00000001071ad881 _ZN7CorUnix10CPalThread11ThreadEntryEPv + 417
  29  libsystem_pthread.dylib             0x00007ff8178564e1 _pthread_start + 125
  30  libsystem_pthread.dylib             0x00007ff817851f6b thread_start + 15!!! IMPORTANT: Please report this at https://github.com/realm/realm-core/issues/new/choose
”
ExceptionPath = Root
ClassName = JetBrains.Util.LoggerException
HResult = COR_E_APPLICATION=80131600
StackTraceString = “
  at JetBrains.Util.ILoggerEx.LogMessage(ILogger this, LoggingLevel level, String message)
     at JetBrains.ReSharper.UnitTestFramework.Execution.TestRunner.TestRunnerRunStrategy.AnalyseProcessCrash(ProcessExitedUnexpectedlyException ex, IUnitTestRun run)
     at JetBrains.ReSharper.UnitTestFramework.Execution.TestRunner.TestRunnerRunStrategy.Run(IUnitTestRun run)
     at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
     at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.AsyncStateMachineBox`1.MoveNext(Thread threadPoolThread)
     at System.Threading.Tasks.AwaitTaskContinuation.RunOrScheduleAction(IAsyncStateMachineBox box, Boolean allowInlining)
     at System.Threading.Tasks.Task.RunContinuations(Object continuationObject)
     at System.Threading.Tasks.Task.FinishSlow(Boolean userDelegateExecute)
     at System.Threading.Tasks.Task.TrySetException(Object exceptionObject)
     at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.SetException(Exception exception, Task`1& taskField)
     at JetBrains.ReSharper.UnitTestFramework.Execution.TestRunner.TestRunnerAgentManager.TestRunnerExecutionAgent.RunTests(CancellationToken cancelCt, CancellationToken abortCt)
     at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
     at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.AsyncStateMachineBox`1.MoveNext(Thread threadPoolThread)
     at System.Threading.Tasks.AwaitTaskContinuation.RunOrScheduleAction(IAsyncStateMachineBox box, Boolean allowInlining)
     at System.Threading.Tasks.Task.RunContinuations(Object continuationObject)
     at System.Threading.Tasks.Task.FinishSlow(Boolean userDelegateExecute)
     at System.Threading.Tasks.Task.TrySetException(Object exceptionObject)
     at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.SetException(Exception exception, Task`1& taskField)
     at JetBrains.ReSharper.UnitTestFramework.Execution.TestRunner.TestRunnerAgentManager.BoundTestRunnerAgent.RunSafe(Func`1 taskFactory, CancellationToken operationAborted)
     at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
     at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.AsyncStateMachineBox`1.MoveNext(Thread threadPoolThread)
     at System.Threading.Tasks.AwaitTaskContinuation.RunOrScheduleAction(IAsyncStateMachineBox box, Boolean allowInlining)
     at System.Threading.Tasks.Task.RunContinuations(Object continuationObject)
     at System.Threading.Tasks.Task.FinishSlow(Boolean userDelegateExecute)
     at System.Threading.Tasks.Task.TrySetException(Object exceptionObject)
     at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.SetException(Exception exception, Task`1& taskField)
     at JetBrains.ReSharper.UnitTestFramework.Common.Extensions.TaskExtensions.ThrowIf[TResult,TException](Task`1 task, WaitHandle waitHandle, Func`1 exceptionFactory)
     at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
     at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.AsyncStateMachineBox`1.MoveNext(Thread threadPoolThread)
     at System.Threading.Tasks.AwaitTaskContinuation.RunOrScheduleAction(IAsyncStateMachineBox box, Boolean allowInlining)
     at System.Threading.Tasks.Task.RunContinuations(Object continuationObject)
     at System.Threading.Tasks.Task`1.TrySetResult(TResult result)
     at System.Threading.Tasks.Task.TwoTaskWhenAnyPromise`1.Invoke(Task completingTask)
     at System.Threading.Tasks.Task.RunContinuations(Object continuationObject)
     at System.Threading.Tasks.Task`1.TrySetResult(TResult result)
     at JetBrains.ReSharper.UnitTestFramework.Common.Extensions.TaskExtensions.<>c__14`2.<ThrowIf>b__14_0(Object s, Boolean timedOut)
     at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
     at System.Threading.PortableThreadPool.CompleteWait(RegisteredWaitHandle handle, Boolean timedOut)
     at System.Threading.ThreadPoolWorkQueue.Dispatch()
     at System.Threading.PortableThreadPool.WorkerThread.WorkerThreadStart()
     at System.Threading.Thread.StartCallback()
”

17:17:32.033 |T| Run: d2a590f5-2c5f-4887-a352-e9820aa3196d >> RunHostController.CleanupAfterRun
17:17:32.033 |T| Run: d2a590f5-2c5f-4887-a352-e9820aa3196d << RunHostController.CleanupAfterRun
17:17:32.034 |I| Run: d2a590f5-2c5f-4887-a352-e9820aa3196d - Finished
17:17:32.034 |I| RunTestsStage completed
17:17:32.034 |I| FinishLaunchStage started
17:17:32.034 |I| Finishing launch
17:17:32.034 |W| Element <TestProject1.UnitTest1.Test> was left running after its run completion.
17:17:32.035 |W| Element <TestProject1.UnitTest1.TestAsync> was left pending after its run completion.
17:17:32.035 |W| Element <TestProject1.UnitTest1> was left running after its run completion.
17:17:32.035 |I| Launch is finished
17:17:32.036 |I| FinishLaunchStage completed
17:17:32.036 |T| Launch: ed7c88be-23a8-4cbf-b090-d086a01e71db >> RunHostController.BeforeLaunchFinished
17:17:32.036 |T| Launch: ed7c88be-23a8-4cbf-b090-d086a01e71db << RunHostController.BeforeLaunchFinished
```

**Comment by papafe** [comment]
@PawKanarek Thanks a lot for your report. I've managed to reproduce the issue locally and we will investigate.

**Comment by PawKanarek** [comment]
@papafe thanks! One more info that can help you. Tests started to fail when I implemented disposing realm instances (earlier I wasn't closing them, but I saw some OOM exceptions on my users devices ), so since I started disposing realms that error hppens randomly.

**Comment by peppy** [comment]
I was a bit curious as to what was going wrong here and can confirm reproducing the issue with this even-more minimal code:

```csharp
using Realms;

namespace TestProject1
{
    public class UnitTest1
    {
        [Fact]
        public void Test()
        {
            var realmConfig = new RealmConfiguration("test.realm");
            File.Delete(realmConfig.DatabasePath);

            for (int i = 0; i < 1000; i++)
            {
                using var realm = Realm.GetInstance(realmConfig);

                realm.Write(() => realm.Add(new RealmObj()));
            }
        }
    }

    public class RealmObj : RealmObject
    {
        [PrimaryKey] private Guid Id { get; set; } = Guid.NewGuid();
    }
}
```

Figured it may be helpful (and rules out `BeginWrite` or the extra instance creation being related). Increasing the number of write operations also increases the chances of a failure - with 1000 it will fail 100% of the time for me.

**Comment by papafe** [comment]
@peppy Can I ask you how did you run the test on your example? On Windows the test is passing 100% of the times, while on macOS I'm getting an error related to the test process crashing.

**Comment by peppy** [comment]
I've only tested on macOS, where it does fail as you say.

[RealmXUnitTest.zip](https://github.com/realm/realm-dotnet/files/9010081/RealmXUnitTest.zip)

```csharp
~/Projects
❯ cd RealmXUnitTest/

~/Projects/RealmXUnitTest
❯ dotnet test
  Determining projects to restore...
  All projects are up-to-date for restore.
  RealmXUnitTest -> /Users/dean/Projects/RealmXUnitTest/RealmXUnitTest/bin/Debug/net6.0/RealmXUnitTest.dll
Test run for /Users/dean/Projects/RealmXUnitTest/RealmXUnitTest/bin/Debug/net6.0/RealmXUnitTest.dll (.NETCoreApp,Version=v6.0)
Microsoft (R) Test Execution Command Line Tool Version 17.2.0 (x64)
Copyright (c) Microsoft Corporation.  All rights reserved.

Starting test execution, please wait...
A total of 1 test files matched the specified pattern.

Passed!  - Failed:     0, Passed:     1, Skipped:     0, Total:     1, Duration: < 1 ms - /Users/dean/Projects/RealmXUnitTest/RealmXUnitTest/bin/Debug/net6.0/RealmXUnitTest.dll (net6.0)

~/Projects/RealmXUnitTest took 10s
❯ dotnet test
  Determining projects to restore...
  All projects are up-to-date for restore.
  RealmXUnitTest -> /Users/dean/Projects/RealmXUnitTest/RealmXUnitTest/bin/Debug/net6.0/RealmXUnitTest.dll
Test run for /Users/dean/Projects/RealmXUnitTest/RealmXUnitTest/bin/Debug/net6.0/RealmXUnitTest.dll (.NETCoreApp,Version=v6.0)
Microsoft (R) Test Execution Command Line Tool Version 17.2.0 (x64)
Copyright (c) Microsoft Corporation.  All rights reserved.

Starting test execution, please wait...
A total of 1 test files matched the specified pattern.
The active test run was aborted. Reason: Test host process crashed

Test Run Aborted with error System.Exception: One or more errors occurred.
 ---> System.Exception: Unable to read beyond the end of the stream.
   at System.IO.BinaryReader.Read7BitEncodedInt()
   at System.IO.BinaryReader.ReadString()
   at Microsoft.VisualStudio.TestPlatform.CommunicationUtilities.LengthPrefixCommunicationChannel.NotifyDataAvailable()
   at Microsoft.VisualStudio.TestPlatform.CommunicationUtilities.TcpClientExtensions.MessageLoopAsync(TcpClient client, ICommunicationChannel channel, Action`1 errorHandler, CancellationToken cancellationToken)
   --- End of inner exception stack trace ---.

~/Projects/RealmXUnitTest took 6s
❯

```

**Comment by papafe** [comment]
@peppy I get a similar error. That is different from the original one, so probably it's caused by something else.

**Comment by papafe** [comment]
@PawKanarek I think this is another case of XUnit's synchronization context behaving in an unexpected way, as in #2642.
As suggested there, if you install the `Nito.AsyncEx.Context` package and wrap your code into `AsyncContext.Run`, then the test is passing all the time. I've also tested it with `NUnit`, and it also works fine there. 

Additionally I could reproduce it with older versions of Realm (tested `10.10.0`), but not at all on windows.

**Comment by PawKanarek** [comment]
Thank you for investigating! I'll add AsyncEx.Context everywhere and i'll see whats happen :)

**Comment by PawKanarek** [comment]
i've added everywhere `AsyncContext.Run`. And in example provided in this ticket it works like a charm. 100% rate of success. 

But in my original project unit tests still fail even after adding `AsyncContext.Run` to every test...
@papafe can you help me with analysing this output from failed test? 

```

18:27:59.192 |I| Launch created for session 'All tests from <Tests>\<InsysGo.Sdk.Portable.UnitTest>' (1d7eeaca-b6c2-4925-b660-2bcae8a34b34)
18:27:59.193 |T| Launch: dfa28b25-d50e-4242-b5cb-362b71ada556 >> RunHostController.AfterLaunchStarted
18:27:59.193 |T| Launch: dfa28b25-d50e-4242-b5cb-362b71ada556 << RunHostController.AfterLaunchStarted
18:27:59.193 |I| BuildStage started
18:27:59.197 |I| Build policy: Automatic
18:28:01.727 |I| 4 projects built
18:28:01.727 |V|     InsysGo.Sdk.Abstractions
    InsysGo.Sdk.Portable
    InsysGo.Sdk.Strings
    InsysGo.Sdk.Portable.UnitTest
18:28:01.727 |I| BuildStage completed
18:28:01.727 |I| RefreshProjectPropertiesStage started
18:28:01.727 |I| RefreshProjectPropertiesStage completed
18:28:01.727 |I| DiscoveryStage started
18:28:01.728 |I| Test exploration required for 1 projects
18:28:01.728 |V|     InsysGo.Sdk.Portable.UnitTest
18:28:01.729 |I| DiscoveryStage completed
18:28:01.729 |I| BuildPipelineStage started
18:28:01.729 |I| BuildPipelineStage completed
18:28:01.729 |I| ApplyRulesStage started
18:28:01.730 |T| Removing elements marked with ignored categories, if any...
18:28:01.730 |T| Removing empty parent elements, if any...
18:28:01.730 |V| Running elements from criterion: {ConjunctiveCriterion: {ProjectCriterion: /Users/pkanarek/Git/insysgo-sdk-mobile/src/Tests/InsysGo.Sdk.Portable.UnitTest}, {NamespaceCriterion: InsysGo.Sdk.Portable.UnitTest.IntegrationTests.TvGuideRepositoryTests}}
18:28:01.730 |I| Got 10 elements (0 explicit) to run
18:28:01.730 |V| Marking 10 elements as Pending
18:28:01.735 |V| Marking elements as Pending completed
18:28:01.735 |V| Distributing 10 elements into runs
18:28:01.736 |V| Distributing elements completed
18:28:01.736 |I| Created 1 unit-test runs
18:28:01.736 |I| ApplyRulesStage completed
18:28:01.736 |I| RunTestsStage started
18:28:01.736 |I| Running up to 10 unit-test runs in parallel
18:28:01.736 |I| Run: 2176fbb3-f668-4daf-b62b-f8af0a3cd3e5 - Starting
18:28:01.736 |V|     Provider: xUnit.net
    Target Framework: netcoreapp3.0
    Strategy: XUnitTestRunnerRunStrategy
    Runtime Environment: TestRunnerRuntimeEnvironment
  Host: JetBrains.ReSharper.UnitTestFramework.Execution.TestRunner.DefaultTestRunnerHost
  Project: InsysGo.Sdk.Portable.UnitTest
  TargetPlatform: X64
  TargetFrameworkId: .NETCoreApp,Version=v3.0
  RunAsTargetFrameworkId: .NETCoreApp,Version=v3.0
  IsUnmanaged: False
18:28:01.736 |T| Run: 2176fbb3-f668-4daf-b62b-f8af0a3cd3e5 >> RunHostController.PrepareForRun
18:28:01.736 |T| Run: 2176fbb3-f668-4daf-b62b-f8af0a3cd3e5 << RunHostController.PrepareForRun
18:28:01.740 |V| RunSettings: 
<RunSettings><RunConfiguration><ResultsDirectory>/Users/pkanarek/Git/insysgo-sdk-mobile/src/TestResults</ResultsDirectory><SolutionDirectory>/Users/pkanarek/Git/insysgo-sdk-mobile/src</SolutionDirectory><TargetPlatform>X64</TargetPlatform></RunConfiguration></RunSettings>
18:28:01.741 |V| [ProcessStartInfoPatchResult]
  Original request: DotNetCoreRequest
  Success: True
  ErrorMessage: <Empty>
  Original StartInfo: FileName = 'exec', Arguments = '  --runtimeConfig /Users/pkanarek/Git/insysgo-sdk-mobile/src/Tests/InsysGo.Sdk.Portable.UnitTest/bin/Debug/netcoreapp3.0/InsysGo.Sdk.Portable.UnitTest.runtimeconfig.json --depsfile /Users/pkanarek/Git/insysgo-sdk-mobile/src/Tests/InsysGo.Sdk.Portable.UnitTest/bin/Debug/netcoreapp3.0/InsysGo.Sdk.Portable.UnitTest.deps.json  --additional-deps /Applications/Rider.app/Contents/lib/ReSharperHost/TestRunner/netcoreapp3.0/ReSharperTestRunner.deps.json --additionalprobingpath /Applications/Rider.app/Contents/lib/ReSharperHost/TestRunner/netcoreapp3.0 /Applications/Rider.app/Contents/lib/ReSharperHost/TestRunner/netcoreapp3.0/ReSharperTestRunner.dll --parentProcessId 30970 -p 59198 -r f2c22098-e90d-4a3f-a589-7598e8ffd0e7', WorkDir = '/Users/pkanarek/Git/insysgo-sdk-mobile/src/Tests/InsysGo.Sdk.Portable.UnitTest/bin/Debug/netcoreapp3.0'
  Patched  StartInfo: FileName = '/usr/local/share/dotnet/dotnet', Arguments = '"exec"   --runtimeConfig /Users/pkanarek/Git/insysgo-sdk-mobile/src/Tests/InsysGo.Sdk.Portable.UnitTest/bin/Debug/netcoreapp3.0/InsysGo.Sdk.Portable.UnitTest.runtimeconfig.json --depsfile /Users/pkanarek/Git/insysgo-sdk-mobile/src/Tests/InsysGo.Sdk.Portable.UnitTest/bin/Debug/netcoreapp3.0/InsysGo.Sdk.Portable.UnitTest.deps.json  --additional-deps /Applications/Rider.app/Contents/lib/ReSharperHost/TestRunner/netcoreapp3.0/ReSharperTestRunner.deps.json --additionalprobingpath /Applications/Rider.app/Contents/lib/ReSharperHost/TestRunner/netcoreapp3.0 /Applications/Rider.app/Contents/lib/ReSharperHost/TestRunner/netcoreapp3.0/ReSharperTestRunner.dll --parentProcessId 30970 -p 59198 -r f2c22098-e90d-4a3f-a589-7598e8ffd0e7', WorkDir = '/Users/pkanarek/Git/insysgo-sdk-mobile/src/Tests/InsysGo.Sdk.Portable.UnitTest/bin/Debug/netcoreapp3.0'

18:28:01.741 |I| Starting process: /usr/local/share/dotnet/dotnet "exec"   --runtimeConfig /Users/pkanarek/Git/insysgo-sdk-mobile/src/Tests/InsysGo.Sdk.Portable.UnitTest/bin/Debug/netcoreapp3.0/InsysGo.Sdk.Portable.UnitTest.runtimeconfig.json --depsfile /Users/pkanarek/Git/insysgo-sdk-mobile/src/Tests/InsysGo.Sdk.Portable.UnitTest/bin/Debug/netcoreapp3.0/InsysGo.Sdk.Portable.UnitTest.deps.json  --additional-deps /Applications/Rider.app/Contents/lib/ReSharperHost/TestRunner/netcoreapp3.0/ReSharperTestRunner.deps.json --additionalprobingpath /Applications/Rider.app/Contents/lib/ReSharperHost/TestRunner/netcoreapp3.0 /Applications/Rider.app/Contents/lib/ReSharperHost/TestRunner/netcoreapp3.0/ReSharperTestRunner.dll --parentProcessId 30970 -p 59198 -r f2c22098-e90d-4a3f-a589-7598e8ffd0e7
18:28:01.783 |I| Started process with pid 32594
18:28:02.139 |V| Test runner agent <f2c22098-e90d-4a3f-a589-7598e8ffd0e7> connected
18:28:02.141 |V| Registering remote tasks...
18:28:02.142 |V| Registering message handlers from 'JetBrains.ReSharper.UnitTestFramework.Execution.TestRunner.DefaultTestRunnerHost'...
18:28:02.145 |V| Registering message handlers from 'JetBrains.ReSharper.UnitTesting.MSTest.Provider.DotNetCore.Proprietary.MsTestTestRunnerOrchestrator'...
18:28:02.145 |V| Registering message handlers from 'JetBrains.ReSharper.UnitTesting.VsTest.Provider.VsTestTestRunnerOrchestrator'...
18:28:02.145 |V| Registering message handlers from 'JetBrains.ReSharper.UnitTestProvider.nUnit.v30.NUnitTestRunnerOrchestrator'...
18:28:02.145 |V| Registering message handlers from 'JetBrains.ReSharper.UnitTestProvider.nUnit.v26.NUnitTestRunnerOrchestrator'...
18:28:02.145 |V| Registering message handlers from 'JetBrains.ReSharper.UnitTestProvider.Xunit.DotNetCore.Proprietary.XUnitTestRunnerOrchestrator'...
18:28:02.158 |V| TestRunner: JetBrains.ReSharper.TestRunner.Implementation.TestAdapterLoadContextFactoryFactory Registered protocol entities: 8 scalars and 1 serializers 
18:28:02.171 |I| TestRunner: JetBrains.ReSharper.TestRunner.Isolation.ShadowCopyBehavior+NoShadowCopyBehavior Shadow-copy will be skipped as it is disabled 
18:28:02.173 |V| TestRunner: JetBrains.ReSharper.TestRunner.Implementation.DefaultTestAdapterLoadContextFactory Initializing test adapter load context: NoIsolation = True 
18:28:02.175 |V| TestRunner: AssemblyLoader(Application) Added probing path '/Users/pkanarek/Git/insysgo-sdk-mobile/src/Tests/InsysGo.Sdk.Portable.UnitTest/bin/Debug/netcoreapp3.0' 
18:28:02.181 |V| TestRunner: AssemblyLoader(Application) Added resolver relative to '/Users/pkanarek/Git/insysgo-sdk-mobile/src/Tests/InsysGo.Sdk.Portable.UnitTest/bin/Debug/netcoreapp3.0/InsysGo.Sdk.Portable.UnitTest.dll' 
18:28:02.182 |V| TestRunner: AssemblyLoader(Application) Added probing path '/Applications/Rider.app/Contents/lib/ReSharperHost/TestRunner/Adapters/XUnit/netstandard2.0' 
18:28:02.182 |V| TestRunner: AssemblyLoader(Application) Added resolver relative to '/Applications/Rider.app/Contents/lib/ReSharperHost/TestRunner/Adapters/XUnit/netstandard2.0/JetBrains.ReSharper.TestRunner.Adapters.XUnit.dll' 
18:28:02.184 |V| TestRunner: JetBrains.ReSharper.TestRunner.Implementation.TestAdapterLoadContextWrapper Initialized test executor: JetBrains.ReSharper.TestRunner.Adapters.XUnit.XUnitRunner, JetBrains.ReSharper.TestRunner.Adapters.XUnit, Version=2.9.1.29, Culture=neutral, PublicKeyToken=5c492ec4f3eccde3 
18:28:02.199 |I| TestRunner: JetBrains.ReSharper.TestRunner.Adapters.XUnit.XUnitRunner Execution started 
18:28:02.203 |I| TestRunner: JetBrains.ReSharper.TestRunner.Adapters.XUnit.Discoverer Discovering tests from /Users/pkanarek/Git/insysgo-sdk-mobile/src/Tests/InsysGo.Sdk.Portable.UnitTest/bin/Debug/netcoreapp3.0/InsysGo.Sdk.Portable.UnitTest.dll 
18:28:02.264 |V| TestRunner: JetBrains.ReSharper.TestRunner.Adapters.XUnit.XUnitRunner Non-serializable data ('System.Object[]') found for 'InsysGo.Sdk.Portable.UnitTest.Tests.PluralTextService.PolishPluralTextServiceTest.PluralizeText_AllFormsProvided_ReturnsCorrectForm'; falling back to single test case. 
18:28:02.267 |V| TestRunner: JetBrains.ReSharper.TestRunner.Adapters.XUnit.XUnitRunner Non-serializable data ('System.Object[]') found for 'InsysGo.Sdk.Portable.UnitTest.Tests.PluralTextService.PolishPluralTextServiceTest.PluralizeText_SecondPluralFormMissing_ReturnsFallbackForm'; falling back to single test case. 
18:28:02.269 |V| TestRunner: JetBrains.ReSharper.TestRunner.Adapters.XUnit.XUnitRunner Non-serializable data ('System.Object[]') found for 'InsysGo.Sdk.Portable.UnitTest.Tests.PluralTextService.PolishPluralTextServiceTest.PluralizeText_OnlySingularFormProvided_ReturnsFallbackForm'; falling back to single test case. 
18:28:02.276 |V| TestRunner: JetBrains.ReSharper.TestRunner.Adapters.XUnit.XUnitRunner Non-serializable data ('System.Object[]') found for 'InsysGo.Sdk.Portable.UnitTest.Tests.ColorTests.ColorTest.ColorTests_Relative'; falling back to single test case. 
18:28:02.352 |V| TestRunner: JetBrains.ReSharper.TestRunner.Adapters.XUnit.Discoverer Sending 465 discovered elements to server... 
18:28:02.357 |V| Run: 2176fbb3-f668-4daf-b62b-f8af0a3cd3e5 - Discovery result processing started
18:28:02.362 |V| Run: 2176fbb3-f668-4daf-b62b-f8af0a3cd3e5 - Discovery result processing finished (+0 -0 ~10)
18:28:02.362 |I| TestRunner: JetBrains.ReSharper.TestRunner.Adapters.XUnit.Discoverer Discovery completed 
18:28:24.278 |I| Process /usr/local/share/dotnet/dotnet:32594 has exited with code (134)
18:28:24.278 |I| Output stream: libc++abi: terminating with uncaught exception of type realm::LogicError: Wrong transactional state (no active transaction, wrong type of transaction, or transaction already in progress)
Exception backtrace:
0   librealm-wrappers.dylib             0x00000001130e1513 _ZN5realm2DB5closeEb + 243
1   librealm-wrappers.dylib             0x00000001130ecd3b _ZNSt3__120__shared_ptr_emplaceIN12_GLOBAL__N_16DBInitENS_9allocatorIS2_EEE16__on_zero_sharedEv + 27
2   librealm-wrappers.dylib             0x0000000112e58b89 _ZNSt3__120__shared_ptr_emplaceIN5realm5_impl16RealmCoordinatorENS_9allocatorIS3_EEE16__on_zero_sharedEv + 1529
3   librealm-wrappers.dylib             0x0000000112ee3f53 _ZN5realm5Realm5closeEv + 355
4   librealm-wrappers.dylib             0x0000000112de6547 shared_realm_close_realm + 23
5   ???                                 0x00000001187b557d 0x0 + 4705703293
6   ???                                 0x000000011a0e5411 0x0 + 4732113937
7   ???                                 0x000000011a0e5305 0x0 + 4732113669
8   ???                                 0x000000011884a093 0x0 + 4706312339
9   ???                                 0x0000000118849ff1 0x0 + 4706312177
10  ???                                 0x00000001187c07ff 0x0 + 4705748991
11  ???                                 0x00000001188519bd 0x0 + 4706343357
12  ???                                 0x000000011a0fd15e 0x0 + 4732211550
13  ???                                 0x00000001188be8c6 0x0 + 4706789574
14  ???                                 0x00000001188be220 0x0 + 4706787872
15  ???                                 0x000000011a0c7c14 0x0 + 4731993108
16  ???                                 0x000000011a0ac903 0x0 + 4731881731
17  ???                                 0x00000001187c47bc 0x0 + 4705765308
18  ???                                 0x0000000115d86938 0x0 + 4661471544
19  ???                                 0x00000001188519bd 0x0 + 4706343357
20  ???                                 0x0000000115d86a3d 0x0 + 4661471805
21  ???                                 0x0000000115d869cd 0x0 + 4661471693
22  ???                                 0x00000001188be220 0x0 + 4706787872
23  ???                                 0x000000011a0c7c14 0x0 + 4731993108
24  ???                                 0x000000011a0ac903 0x0 + 4731881731
25  ???                                 0x00000001187e9201 0x0 + 4705915393
26  ???                                 0x0000000115d86938 0x0 + 4661471544
27  ???                                 0x000000011a0e1cd1 0x0 + 4732099793
28  ???                                 0x0000000115d86a63 0x0 + 4661471843
29  ???                                 0x0000000115d869ab 0x0 + 4661471659
30  libcoreclr.dylib                    0x000000010f74903d CallDescrWorkerInternal + 124
31  libcoreclr.dylib                    0x000000010f5ab81a _ZN18MethodDescCallSite16CallTargetWorkerEPKmPmi + 1530
32  libcoreclr.dylib                    0x000000010f5c5030 _Z32QueueUserWorkItemManagedCallbackPv + 160
33  libcoreclr.dylib                    0x000000010f56f5f8 _ZL31ManagedThreadBase_DispatchOuterP22ManagedThreadCallState + 296
34  libcoreclr.dylib                    0x000000010f56fc00 _ZN17ManagedThreadBase10ThreadPoolEPFvPvES0_ + 32
35  libcoreclr.dylib                    0x000000010f5667fb _ZN26ManagedPerAppDomainTPCount16DispatchWorkItemEPbS0_ + 283
36  libcoreclr.dylib                    0x000000010f5925ea _ZN13ThreadpoolMgr17WorkerThreadStartEPv + 1066
37  libcoreclr.dylib                    0x000000010f3f8881 _ZN7CorUnix10CPalThread11ThreadEntryEPv + 417
38  libsystem_pthread.dylib             0x00007ff8171504e1 _pthread_start + 125
39  libsystem_pthread.dylib             0x00007ff81714bf6b thread_start + 15


18:28:24.380 |E| Process /usr/local/share/dotnet/dotnet:32594 exited with code '134': libc++abi: terminating with uncaught exception of type realm::LogicError: Wrong transactional state (no active transaction, wrong type of transaction, or transaction already in progress) Exception backtrace: 0 librealm-wrappers.dylib 0x00000001130e1513 _ZN5realm2DB5closeEb + 243 1 librealm-wrappers.dylib 0x00000001130ecd3b _ZNSt3__120__shared_ptr_emplaceIN12_GLOBAL__N_16DBInitENS_9allocatorIS2_EEE16__on_zero_sharedEv + 27 2 librealm-wrappers.dylib 0x0000000112e58b89 _ZNSt3__120__shared_ptr_emplaceIN5realm5_impl16RealmCoordinatorENS_9allocatorIS3_EEE16__on_zero_sharedEv + 1529 3 librealm-wrappers.dylib 0x0000000112ee3f53 _ZN5realm5Realm5closeEv + 355 4 librealm-wrappers.dylib 0x0000000112de6547 shared_realm_close_realm + 23 5 ??? 0x00000001187b557d 0x0 + 4705703293 6 ??? 0x000000011a0e5411 0x0 + 4732113937 7 ??? 0x000000011a0e5305 0x0 + 4732113669 8 ??? 0x000000011884a093 0x0 + 4706312339 9 ??? 0x0000000118849ff1 0x0 + 4706312177 10 ??? 0x00000001187c07ff 0x0 + 4705748991 11 ??? 0x00000001188519bd 0x0 + 4706343357 12 ??? 0x000000011a0fd15e 0x0 + 4732211550 13 ??? 0x00000001188be8c6 0x0 + 4706789574 14 ??? 0x00000001188be220 0x0 + 4706787872 15 ??? 0x000000011a0c7c14 0x0 + 4731993108 16 ??? 0x000000011a0ac903 0x0 + 4731881731 17 ??? 0x00000001187c47bc 0x0 + 4705765308 18 ??? 0x0000000115d86938 0x0 + 4661471544 19 ??? 0x00000001188519bd 0x0 + 4706343357 20 ??? 0x0000000115d86a3d 0x0 + 4661471805 21 ??? 0x0000000115d869cd 0x0 + 4661471693 22 ??? 0x00000001188be220 0x0 + 4706787872 23 ??? 0x000000011a0c7c14 0x0 + 4731993108 24 ??? 0x000000011a0ac903 0x0 + 4731881731 25 ??? 0x00000001187e9201 0x0 + 4705915393 26 ??? 0x0000000115d86938 0x0 + 4661471544 27 ??? 0x000000011a0e1cd1 0x0 + 4732099793 28 ??? 0x0000000115d86a63 0x0 + 4661471843 29 ??? 0x0000000115d869ab 0x0 + 4661471659 30 libcoreclr.dylib 0x000000010f74903d CallDescrWorkerInternal + 124 31 libcoreclr.dylib 0x000000010f5ab81a _ZN18MethodDescCallSite16CallTargetWorkerEPKmPmi + 1530 32 libcoreclr.dylib 0x000000010f5c5030 _Z32QueueUserWorkItemManagedCallbackPv + 160 33 libcoreclr.dylib 0x000000010f56f5f8 _ZL31ManagedThreadBase_DispatchOuterP22ManagedThreadCallState + 296 34 libcoreclr.dylib 0x000000010f56fc00 _ZN17ManagedThreadBase10ThreadPoolEPFvPvES0_ + 32 35 libcoreclr.dylib 0x000000010f5667fb _ZN26ManagedPerAppDomainTPCount16DispatchWorkItemEPbS0_ + 283 36 libcoreclr.dylib 0x000000010f5925ea _ZN13ThreadpoolMgr17WorkerThreadStartEPv + 1066 37 libcoreclr.dylib 0x000000010f3f8881 _ZN7CorUnix10CPalThread11ThreadEntryEPv + 417 38 libsystem_pthread.dylib 0x00007ff8171504e1 _pthread_start + 125 39 libsystem_pthread.dylib 0x00007ff81714bf6b thread_start + 15

--- EXCEPTION #1/1 [LoggerException]
Message = “
  Process /usr/local/share/dotnet/dotnet:32594 exited with code '134':
  libc++abi: terminating with uncaught exception of type realm::LogicError: Wrong transactional state (no active transaction, wrong type of transaction, or transaction already in progress)
  Exception backtrace:
  0   librealm-wrappers.dylib             0x00000001130e1513 _ZN5realm2DB5closeEb + 243
  1   librealm-wrappers.dylib             0x00000001130ecd3b _ZNSt3__120__shared_ptr_emplaceIN12_GLOBAL__N_16DBInitENS_9allocatorIS2_EEE16__on_zero_sharedEv + 27
  2   librealm-wrappers.dylib             0x0000000112e58b89 _ZNSt3__120__shared_ptr_emplaceIN5realm5_impl16RealmCoordinatorENS_9allocatorIS3_EEE16__on_zero_sharedEv + 1529
  3   librealm-wrappers.dylib             0x0000000112ee3f53 _ZN5realm5Realm5closeEv + 355
  4   librealm-wrappers.dylib             0x0000000112de6547 shared_realm_close_realm + 23
  5   ???                                 0x00000001187b557d 0x0 + 4705703293
  6   ???                                 0x000000011a0e5411 0x0 + 4732113937
  7   ???                                 0x000000011a0e5305 0x0 + 4732113669
  8   ???                                 0x000000011884a093 0x0 + 4706312339
  9   ???                                 0x0000000118849ff1 0x0 + 4706312177
  10  ???                                 0x00000001187c07ff 0x0 + 4705748991
  11  ???                                 0x00000001188519bd 0x0 + 4706343357
  12  ???                                 0x000000011a0fd15e 0x0 + 4732211550
  13  ???                                 0x00000001188be8c6 0x0 + 4706789574
  14  ???                                 0x00000001188be220 0x0 + 4706787872
  15  ???                                 0x000000011a0c7c14 0x0 + 4731993108
  16  ???                                 0x000000011a0ac903 0x0 + 4731881731
  17  ???                                 0x00000001187c47bc 0x0 + 4705765308
  18  ???                                 0x0000000115d86938 0x0 + 4661471544
  19  ???                                 0x00000001188519bd 0x0 + 4706343357
  20  ???                                 0x0000000115d86a3d 0x0 + 4661471805
  21  ???                                 0x0000000115d869cd 0x0 + 4661471693
  22  ???                                 0x00000001188be220 0x0 + 4706787872
  23  ???                                 0x000000011a0c7c14 0x0 + 4731993108
  24  ???                                 0x000000011a0ac903 0x0 + 4731881731
  25  ???                                 0x00000001187e9201 0x0 + 4705915393
  26  ???                                 0x0000000115d86938 0x0 + 4661471544
  27  ???                                 0x000000011a0e1cd1 0x0 + 4732099793
  28  ???                                 0x0000000115d86a63 0x0 + 4661471843
  29  ???                                 0x0000000115d869ab 0x0 + 4661471659
  30  libcoreclr.dylib                    0x000000010f74903d CallDescrWorkerInternal + 124
  31  libcoreclr.dylib                    0x000000010f5ab81a _ZN18MethodDescCallSite16CallTargetWorkerEPKmPmi + 1530
  32  libcoreclr.dylib                    0x000000010f5c5030 _Z32QueueUserWorkItemManagedCallbackPv + 160
  33  libcoreclr.dylib                    0x000000010f56f5f8 _ZL31ManagedThreadBase_DispatchOuterP22ManagedThreadCallState + 296
  34  libcoreclr.dylib                    0x000000010f56fc00 _ZN17ManagedThreadBase10ThreadPoolEPFvPvES0_ + 32
  35  libcoreclr.dylib                    0x000000010f5667fb _ZN26ManagedPerAppDomainTPCount16DispatchWorkItemEPbS0_ + 283
  36  libcoreclr.dylib                    0x000000010f5925ea _ZN13ThreadpoolMgr17WorkerThreadStartEPv + 1066
  37  libcoreclr.dylib                    0x000000010f3f8881 _ZN7CorUnix10CPalThread11ThreadEntryEPv + 417
  38  libsystem_pthread.dylib             0x00007ff8171504e1 _pthread_start + 125
  39  libsystem_pthread.dylib             0x00007ff81714bf6b thread_start + 15
”
ExceptionPath = Root
ClassName = JetBrains.Util.LoggerException
HResult = COR_E_APPLICATION=80131600
StackTraceString = “
  at JetBrains.Util.ILoggerEx.LogMessage(ILogger this, LoggingLevel level, String message)
     at JetBrains.Util.ILoggerEx.Error(ILogger this, String message)
     at JetBrains.ReSharper.UnitTestFramework.Execution.TestRunner.TestRunnerRunStrategy.AnalyseProcessCrash(ProcessExitedUnexpectedlyException ex, IUnitTestRun run)
     at JetBrains.ReSharper.UnitTestFramework.Execution.TestRunner.TestRunnerRunStrategy.Run(IUnitTestRun run)
     at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
     at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.AsyncStateMachineBox`1.MoveNext(Thread threadPoolThread)
     at System.Threading.Tasks.AwaitTaskContinuation.RunOrScheduleAction(IAsyncStateMachineBox box, Boolean allowInlining)
     at System.Threading.Tasks.Task.RunContinuations(Object continuationObject)
     at System.Threading.Tasks.Task.FinishSlow(Boolean userDelegateExecute)
     at System.Threading.Tasks.Task.TrySetException(Object exceptionObject)
     at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.SetException(Exception exception, Task`1& taskField)
     at JetBrains.ReSharper.UnitTestFramework.Execution.TestRunner.TestRunnerAgentManager.TestRunnerExecutionAgent.RunTests(CancellationToken cancelCt, CancellationToken abortCt)
     at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
     at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.AsyncStateMachineBox`1.MoveNext(Thread threadPoolThread)
     at System.Threading.Tasks.AwaitTaskContinuation.RunOrScheduleAction(IAsyncStateMachineBox box, Boolean allowInlining)
     at System.Threading.Tasks.Task.RunContinuations(Object continuationObject)
     at System.Threading.Tasks.Task.FinishSlow(Boolean userDelegateExecute)
     at System.Threading.Tasks.Task.TrySetException(Object exceptionObject)
     at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.SetException(Exception exception, Task`1& taskField)
     at JetBrains.ReSharper.UnitTestFramework.Execution.TestRunner.TestRunnerAgentManager.BoundTestRunnerAgent.RunSafe(Func`1 taskFactory, CancellationToken operationAborted)
     at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
     at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.AsyncStateMachineBox`1.MoveNext(Thread threadPoolThread)
     at System.Threading.Tasks.AwaitTaskContinuation.RunOrScheduleAction(IAsyncStateMachineBox box, Boolean allowInlining)
     at System.Threading.Tasks.Task.RunContinuations(Object continuationObject)
     at System.Threading.Tasks.Task.FinishSlow(Boolean userDelegateExecute)
     at System.Threading.Tasks.Task.TrySetException(Object exceptionObject)
     at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.SetException(Exception exception, Task`1& taskField)
     at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.SetException(Exception exception)
     at JetBrains.ReSharper.UnitTestFramework.Common.Extensions.TaskExtensions.ThrowIf[TResult,TException](Task`1 task, WaitHandle waitHandle, Func`1 exceptionFactory)
     at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
     at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.AsyncStateMachineBox`1.MoveNext(Thread threadPoolThread)
     at System.Threading.Tasks.AwaitTaskContinuation.RunOrScheduleAction(IAsyncStateMachineBox box, Boolean allowInlining)
     at System.Threading.Tasks.Task.RunContinuations(Object continuationObject)
     at System.Threading.Tasks.Task`1.TrySetResult(TResult result)
     at System.Threading.Tasks.Task.TwoTaskWhenAnyPromise`1.Invoke(Task completingTask)
     at System.Threading.Tasks.Task.RunContinuations(Object continuationObject)
     at System.Threading.Tasks.Task`1.TrySetResult(TResult result)
     at System.Threading.Tasks.TaskCompletionSource`1.TrySetResult(TResult result)
     at JetBrains.ReSharper.UnitTestFramework.Common.Extensions.TaskExtensions.<>c__14`2.<ThrowIf>b__14_0(Object s, Boolean timedOut)
     at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
     at System.Threading.PortableThreadPool.CompleteWait(RegisteredWaitHandle handle, Boolean timedOut)
     at System.Threading.ThreadPoolWorkQueue.Dispatch()
     at System.Threading.PortableThreadPool.WorkerThread.WorkerThreadStart()
     at System.Threading.Thread.StartCallback()
”

18:28:24.380 |T| Run: 2176fbb3-f668-4daf-b62b-f8af0a3cd3e5 >> RunHostController.CleanupAfterRun
18:28:24.380 |T| Run: 2176fbb3-f668-4daf-b62b-f8af0a3cd3e5 << RunHostController.CleanupAfterRun
18:28:24.381 |I| Run: 2176fbb3-f668-4daf-b62b-f8af0a3cd3e5 - Finished
18:28:24.381 |I| RunTestsStage completed
18:28:24.381 |I| FinishLaunchStage started
18:28:24.383 |I| Finishing launch
18:28:24.383 |W| Element <InsysGo.Sdk.Portable.UnitTest.IntegrationTests.TvGuideRepositoryTests.TvGuideRepositoryTests.TvGuideRepository_UpdateTvGuideByHoursAsyncWithNextProgram_ContainsNextPrograms> was left pending after its run completion.
18:28:24.383 |W| Element <InsysGo.Sdk.Portable.UnitTest.IntegrationTests.TvGuideRepositoryTests.TvGuideRepositoryTests.TvGuideRepository_Simulate_OnAppearing_TvGuideByChannels> was left running after its run completion.
18:28:24.384 |W| Element <InsysGo.Sdk.Portable.UnitTest.IntegrationTests.TvGuideRepositoryTests.TvGuideRepositoryTests> was left running after its run completion.
18:28:24.384 |I| Launch is finished
18:28:24.385 |I| FinishLaunchStage completed
18:28:24.385 |T| Launch: dfa28b25-d50e-4242-b5cb-362b71ada556 >> RunHostController.BeforeLaunchFinished
18:28:24.385 |T| Launch: dfa28b25-d50e-4242-b5cb-362b71ada556 << RunHostController.BeforeLaunchFinished


```

**Comment by PawKanarek** [comment]
please look at video: https://github.com/PawKanarek/CustomRendererTest/blob/master/Screen%20Recording%202022-07-04%20at%2018.55.00.mov?raw=true

0:00: all tests passed.
0:13: launching all tests - some fail
0:22 aunching all tests - some fail
0:33 aunching all tests - some fail
1:00 launching 2 tests - success
1:07 launching 1 test - success
1:28 launching 4 tests - success 
1:39 launching 3 tests - success 
1:48 launching all tests - some fail 
**2:42 launching all tests - success** 
2:57 launching all tests - some fail 
and until end of video im presenting some logs from failed test... 


It all started when i've started disposing & closing realm instances :(

**Comment by papafe** [comment]
@PawKanarek the new error that you posted is different from before, so probably it's caused by something different and I think it would make sense to create another ticket. 

It would be really useful if you could send us the project that you're using for testing, or even just a minimal example, so that we reproduce the error and debug locally. We can send you a link where you can upload the project privately if you prefer. Would it be ok for you?

**Comment by PawKanarek** [comment]
@papafe thats too big project to be shared outside my company. And sending just fraction of it won't do much. 

This native crash from the last post happens when i create first realm instance for some tests. Sadly i cannot try catch it, as it just fails without managed c# exception.  Maybe there is something wrong with my idea with creating & deleting .realm files for each new test? 

I'm closing this issue for now, my workaround is to disable unit tests on CI Jenkins... I'll try to recreate behaviour of my app on very simple project, if I succeed, I will create new ticket ;) 

Finally, thank you for your support! Adding AsyncContext increased success rate a bit ❤️

**Comment by papafe** [comment]
@PawKanarek I'm glad at least that helped a little ;)

Regarding your issue.. As far as I remember XUnit by default runs the tests from different collections in parallel. A first thing I would try is to disable it, as there could be some race conditions involved. 

Also, are you sure you're using a different realm file for each test?

**Comment by PawKanarek** [comment]
@papafe  i'll try to start tests in synchronised way ;) And I'll double check if every test uses right instance of .realm. But im like 99% sure that it is using correct file. They are created in /bin/ folder, each test collection creates new dir, and each test case creates own RealmConfiguration. It looks like this: 
![image](https://user-images.githubusercontent.com/15096514/177304413-6d87726c-e2c1-4fcc-955a-757c28a81567.png)
Also I can see it clearly on debugger when test are failing, that they have correct RealmFilePath in Configuration class, and as soon as `Realm.GetInstance(config)` is fired, native crash happens.
